File: backend/models/Task.js
import { DataTypes, Op } from "sequelize";
import sequelize from "../config/database.js";

const Task = sequelize.define(
  "Task",
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true,
    },
    listId: {
      type: DataTypes.UUID,
      allowNull: false,
      references: {
        model: "lists",
        key: "id",
      },
      onUpdate: "CASCADE",
      onDelete: "CASCADE",
    },
    title: {
      type: DataTypes.STRING,
      allowNull: false,
      validate: {
        len: [1, 255],
      },
    },
    description: {
      type: DataTypes.TEXT,
    },
    priority: {
      type: DataTypes.STRING,
      defaultValue: "medium",
      validate: {
        isIn: [["low", "medium", "high", "urgent"]],
      },
    },
    dueDate: {
      type: DataTypes.DATEONLY,
    },
    dueTime: {
      type: DataTypes.TIME,
    },
    completed: {
      type: DataTypes.BOOLEAN,
      defaultValue: false,
    },
    completedAt: {
      type: DataTypes.DATE,
    },
    position: {
      type: DataTypes.INTEGER,
      defaultValue: 0,
    },
    tags: {
      type: DataTypes.ARRAY(DataTypes.STRING),
      defaultValue: [],
    },
    metadata: {
      type: DataTypes.JSON,
      defaultValue: {},
    },
    isRecurring: {
      type: DataTypes.BOOLEAN,
      defaultValue: false,
    },
    recurringPattern: {
      type: DataTypes.JSON,
    },
    parentTaskId: {
      type: DataTypes.UUID,
      references: {
        model: "tasks",
        key: "id",
      },
    },
  },
  {
    tableName: "tasks",
    hooks: {
      beforeUpdate: (task) => {
        if (task.changed("completed") && task.completed) {
          task.completedAt = new Date();
        } else if (task.changed("completed") && !task.completed) {
          task.completedAt = null;
        }
      },
    },
  }
);

export { Task };

File: mobile-app/lib/models/task.dart
class Task {
  final String id;
  final String title;
  final String? description;
  final String listId;
  final String priority;
  final bool completed;
  final String? dueDate;
  final String? dueTime;
  final DateTime createdAt;
  final DateTime updatedAt;

  Task({
    required this.id,
    required this.title,
    this.description,
    required this.listId,
    required this.priority,
    this.completed = false,
    this.dueDate,
    this.dueTime,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Task.fromJson(Map<String, dynamic> json) {
    return Task(
      id: json['id'] ?? '',
      title: json['title'] ?? '',
      description: json['description'],
      listId: json['listId'] ?? '',
      priority: json['priority'] ?? 'medium',
      completed: json['completed'] ?? false,
      dueDate: json['dueDate'],
      dueTime: json['dueTime'],
      createdAt: DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
      updatedAt: DateTime.parse(json['updatedAt'] ?? DateTime.now().toIso8601String()),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'description': description,
      'listId': listId,
      'priority': priority,
      'completed': completed,
      'dueDate': dueDate,
      'dueTime': dueTime,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }
}

class CreateTaskForm {
  final String title;
  final String? description;
  final String listId;
  final String priority;
  final String? dueDate;
  final String? dueTime;

  CreateTaskForm({
    required this.title,
    this.description,
    required this.listId,
    this.priority = 'medium',
    this.dueDate,
    this.dueTime,
  });

  Map<String, dynamic> toJson() {
    return {
      'title': title,
      'description': description,
      'listId': listId,
      'priority': priority,
      'dueDate': dueDate,
      'dueTime': dueTime,
    };
  }
}